process/coordinator/process.go

```diff
@@ -1162,6 +1179,147 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
 	return numCrossShardScCallsOrSpecialTxs
 }
 
+// VerifyCreatedMiniBlocks re-checks gas used and generated fees in the given block
+func (tc *transactionCoordinator) VerifyCreatedMiniBlocks(header data.HeaderHandler, body *block.Body) error {
+	if header.GetEpoch() < tc.blockGasAndFeesReCheckEnableEpoch {
+		return nil
+	}
+
+	mapMiniBlockTypeAllTxs := make(map[block.Type]map[string]data.TransactionHandler)
+	for _, miniBlock := range body.MiniBlocks {
+		_, ok := mapMiniBlockTypeAllTxs[miniBlock.Type]
+		if !ok {
+			mapMiniBlockTypeAllTxs[miniBlock.Type] = tc.GetAllCurrentUsedTxs(miniBlock.Type)
+		}
+	}
+
+	err := tc.verifyGasLimit(body, mapMiniBlockTypeAllTxs)
+	if err != nil {
+		return err
+	}
+
+	err = tc.verifyFees(header, body, mapMiniBlockTypeAllTxs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (tc *transactionCoordinator) verifyGasLimit(
+	body *block.Body,
+	mapMiniBlockTypeAllTxs map[block.Type]map[string]data.TransactionHandler,
+) error {
+	for _, miniBlock := range body.MiniBlocks {
+		isCrossShardMiniBlockFromMe := miniBlock.SenderShardID == tc.shardCoordinator.SelfId() &&
+			miniBlock.ReceiverShardID != tc.shardCoordinator.SelfId()
+		if !isCrossShardMiniBlockFromMe {
+			continue
+		}
+
+		err := tc.checkGasConsumedByMiniBlockInReceiverShard(miniBlock, mapMiniBlockTypeAllTxs[miniBlock.Type])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tc *transactionCoordinator) checkGasConsumedByMiniBlockInReceiverShard(
+	miniBlock *block.MiniBlock,
+	mapHashTx map[string]data.TransactionHandler,
+) error {
+	var err error
+	var gasConsumedByTxInReceiverShard uint64
+	gasConsumedByMiniBlockInReceiverShard := uint64(0)
+
+	for _, txHash := range miniBlock.TxHashes {
+		txHandler, ok := mapHashTx[string(txHash)]
+		if !ok {
+			log.Debug("missing transaction in checkGasConsumedByMiniBlockInReceiverShard ", "type", miniBlock.Type, "txHash", txHash)
+			return process.ErrMissingTransaction
+		}
+
+		_, txTypeDstShard := tc.txTypeHandler.ComputeTransactionType(txHandler)
+		if txTypeDstShard == process.MoveBalance {
+			gasConsumedByTxInReceiverShard = tc.economicsFee.ComputeGasLimit(txHandler)
+		} else {
+			gasConsumedByTxInReceiverShard = txHandler.GetGasLimit()
```

Comment:
Is this correct?
Shouldn't this be txHandler.GetGasLimit() - gas in sender shard?

Reply:
Yep, changed this way :+1: